sdk/trace/provider.go

```diff
@@ -237,14 +237,13 @@ func (p *TracerProvider) ForceFlush(ctx context.Context) error {
 // Shutdown shuts down TracerProvider. All registered span processors are shut down
 // in the order they were registered and any held computational resources are released.
 func (p *TracerProvider) Shutdown(ctx context.Context) error {
-	spss := p.spanProcessors.Load().(spanProcessorStates)
-	if len(spss) == 0 {
```

Comment:
Not strictly necessary because `for` will just make 0 iterations if the slice is empty.

Reply:
No, but it avoids needing to take a lock.  Line 270 (in the original) stores an empty list after the processors have been shutdown so any further invocation will find an empty list and immediately return without waiting on a lock.

Reply:
Lock should be taken anyway to check the `isShutdown` flag.

Reply:
And to set it to `true`

Reply:
The lock-free mechanism for accessing the span processors list is needed as that is also accessed when every span is started or ended.  Because the `Shutdown()` method does not have any work to do if there are no span processors this is an effective short-circuit.  The `isShutdown` flag is only used to short-circuit registering or unregistering span processors.  It will have been set if there were any span processors registered when `Shutdown()` is invoked.  If there were none then the provider had nothing to shutdown and can still have processors registered that may later be shutdown.  

What exactly is this change attempting to do?  Is there a defect it is fixing or is this purely aesthetic?

Reply:
@Aneurysm9 I think it's a bug that these behaviors are not identical:

- Register a processor, shutdown, **cannot** register a processor.
- Shutdown, **can** register a processor.

There is no good reason for this non-determinism. Also, I don't think this change makes anything worse or more complex. On the contrary, it reduces the states the object can be in i.e. it simplifies things.

The second commit also fixes a bug where not all processors are given a chance to perform a cleanup operation.

Is there something that worries you in this PR? Something that you'd like to be done differently?

Reply:
I think this PR needs a better description of what it does and why in the title and changelog.  It also needs tests.  The only reason I see to make the change here is to reconcile the behavior of a provider that has been shut down with processors and one that has been shut down without.  Not testing that means we can wind up in the same situation again later.

As for whether processing should continue after context cancellation I'm not convinced it is correct behavior.  I understand why it might be desirable, but it is also contrary to the expectations of a caller that provides a context for deadline purposes.

Reply:
>  It also needs tests

👍 

> As for whether processing should continue after context cancellation I'm not convinced it is correct behavior. I understand why it might be desirable, but it is also contrary to the expectations of a caller that provides a context for deadline purposes.

For sure it would be better to have it in a separate PR.

Reply:
Added some tests, removed the contentious change. Please let me know if you want it as a separate PR. Let me know if you want me to make further tweaks. Thanks!